### PR TITLE
feat(ctb): Specify L1 auth requirements

### PR DESCRIPTION
**Pull request by Inphi**
**Description**

fixes https://github.com/ethereum-optimism/client-pod/issues/251

Specifies which contracts and their functions require auth.

Every L1 entrypoint is enumerated and annotated to document whether auth is needed. This is enforced by `Authorization_Test` which asserts that there exists an explicit auth spec for every L1 contract.

### CODE REVIEW

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (Inphi) [inline packages/contracts-bedrock/test/Authorization.t.sol:32]: @mds1 For now, each `Spec` documents which methods can only be used by contract owners. What other types of auth are worth documenting? I could see maybe "pauseability" going in here in the future.

Comment (mds1) [reply packages/contracts-bedrock/test/Authorization.t.sol:32]: I like how this file is setup overall, and specifying pausability status makes sense to me also, but cc @maurelian since I know we were discussing that the other day.

The only downside to the current test approach is we don't test that the implementations match the `Spec`'s auth bool. For `Initializeable` this was easy to do because it's small amount of functions and initialization is the first thing checked during execution so calldata doesn't matter. It's a lot harder to do something like in the general case for all functions here, so I think having this act just as a length check + forcing function for the spec is ok (pending a better solution, ref your notion [doc](https://www.notion.so/oplabs/Contract-Specifications-617716b8d19a44b381038d59cbde3c1e?pvs=4))).

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (mds1) [inline packages/contracts-bedrock/test/Authorization.t.sol]: These should have auth of `true`, as should `deleteL2Outputs`, `proposeL2Output`, functions with the `onlyOtherBridge` modifier, and likely some others. Searching for `msg.sender ==` or `== msg.sender` is probably a good way to find most of these. (Totally understand this is a draft so maybe you didn't go through them all yet).

For the ones that are `true` we should also manually a corresponding auth test is present (so far I've confirmed this for `pause` and `unpause` only)

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (mds1) [inline packages/contracts-bedrock/test/Authorization.t.sol]: Might be worth expanding on the comments here with the expected workflow, to say that when new functions are added to the codebase they must also be added here, but it's up to the PR author and reviewers to manually ensure that there are tests for the spec defined here

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (maurelian) [inline packages/contracts-bedrock/test/Authorization.t.sol:241]: Maybe I'm missing something, but shouldn't there be an actual pranked call to each of these functions to validate the auth?

Comment (maurelian) [reply packages/contracts-bedrock/test/Authorization.t.sol:241]: OK, I talked with @mds1 the other day about this, and now understand the value as being that it ensures that a spec exists for all methods.

In the future we can consider extending this to actually test the spec, but for now this seems good.

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (maurelian) [reply packages/contracts-bedrock/test/Authorization.t.sol:32]: +1 to adding pausability on this, but I think it's out of scope for this PR.

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (Inphi) [reply packages/contracts-bedrock/test/Authorization.t.sol:241]: We already have forge tests that do this for every contract.
The aim of these tests is to make sure that there's an auth specification for every interaction with the L1 contracts. It's purely meant for documentation.

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (Inphi) [reply packages/contracts-bedrock/test/Authorization.t.sol:275]: Noted. Foundry is setup to restrict reads to certain directories. Effectively securing all access.

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (mds1) [reply packages/contracts-bedrock/test/Authorization.t.sol:32]: @Inphi What do you think of changing `bool auth` to `Role auth[]` where:
```solidity
enum Role {
  Challenger,
  L1ProxyAdminOwner,
  L2ProxyAdminOwner,
  SystemConfigOwner,
  MintManagerOwner,
  Guardian
}
```
Roles sourced from this [notion doc](https://www.notion.so/oplabs/4ab895de0c36475ab3ef292cf12adc00?v=511aead484f54a06b89fb2672172611d&pvs=4), where MSD Controller is outdated. cc @maurelian for thoughts / confirmation on roles.

If we like this approach for more specificity I'd suggest pulling it into a separate issue to avoid blocking this PR

Comment (maurelian) [reply packages/contracts-bedrock/test/Authorization.t.sol:32]: > If we like this approach for more specificity I'd suggest pulling it into a separate issue to avoid blocking this PR

💯 to doing this. 
Unopinionated about doing it in this PR or separately.

Comment (mds1) [reply packages/contracts-bedrock/test/Authorization.t.sol:32]: Created an issue to track here: https://github.com/ethereum-optimism/client-pod/issues/308

Pulling out to a separate issue seems ideal so we can unblock you to add `paused` status + unblock @Inphi to work on the state diff bug :)

Comment (Inphi) [reply packages/contracts-bedrock/test/Authorization.t.sol:32]: I'll leave this out to be done on another PR. Thanks

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (Inphi) [reply packages/contracts-bedrock/test/Authorization.t.sol:241]: It's intentional. The console.log is really useful to debug missing specs. It's difficult to figure out why the test fails otherwise

File: packages/contracts-bedrock/scripts/Deployer.sol
Comment (Inphi) [reply packages/contracts-bedrock/scripts/Deployer.sol:495]: security risk is minimal. Foundry configuration ensures that only certain paths are read.

**Review by mds1 — APPROVED** [review]
LGTM!

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (maurelian) [inline packages/contracts-bedrock/test/Authorization.t.sol:158]: oh shit, I'm gonna remove this.

**Review by maurelian — APPROVED** [review]

File: packages/contracts-bedrock/test/Authorization.t.sol
Comment (maurelian) [reply packages/contracts-bedrock/test/Authorization.t.sol:158]: Will do in a follow up PR tho.

**Pull request merged**
